This issue is fixed upstream. This patch ports the relevant changes from the 
upstream alpha version.
--- INIT.2012-08-01.old/src/cmd/ksh93/sh/xec.c	2018-03-02 04:54:23.567563320 -0800
+++ INIT.2012-08-01/src/cmd/ksh93/sh/xec.c	2018-03-07 02:23:16.557299090 -0800
@@ -1601,13 +1603,14 @@
 				break;
 			}
 #endif /* SHOPT_COSHELL */
-			if(shp->subshell)
-			{
+			if (shp->subshell) {
 				sh_subtmpfile(shp);
-				if(shp->comsub==1 && !(shp->fdstatus[1]&IONOSEEK))
-					unpipe=iousepipe(shp);
-				if((type&(FAMP|TFORK))==(FAMP|TFORK))
+				if (( type & ( FAMP | TFORK)) == (FAMP | TFORK)) {
+					if (shp->comsub && !(shp->fdstatus[1] & IONOSEEK)) {
+						unpipe = iousepipe(shp);
+					}
 					sh_subfork();
+				}
 			}
 			no_fork = !ntflag && !(type&(FAMP|FPOU)) && !shp->subshell &&
 			    !(shp->st.trapcom[SIGINT] && *shp->st.trapcom[SIGINT]) &&
